### DISCUSSION TITLE
Updating to Craft 4, Freeform no longer has any email templates

### DISCUSSION THREAD

**Discussion by darylknight** [solspace/craft-freeform#374] [Questions & Answers] [answered]
**Describe the bug or issue you're experiencing**
Just updated a site to Craft 4 and tested sending a contact form. I didn't get an email. Went to check the email notifications (which previously had one set up to go to admins, and one to go to the user) and saw this:

![image](https://user-images.githubusercontent.com/1250017/175034152-4d5f9b00-70a9-4bba-8da7-48367b593cc9.png)

It used to look like this in Craft 3:

![image](https://user-images.githubusercontent.com/1250017/175034757-28588776-6649-4ce2-9ee4-1317f8787576.png)

Issues
1. This appears on production, If you click "set directory path in settings" you're taken to an error page because you can't make changes to settings in production
2. Our old email notifications have been removed during the update to Craft 4.
3. After creating a directory and telling Freeform to use it, we can create email notifications locally. It looks like Freeform then tries to record this in project config, and we commit the template which is created automatically to git. However - users can still create notifications on the live server, which would create files in a version controlled folder and either create merge conflicts or be deleted if we were to create notifications of the same name locally then deploy them.

**Expected behavior**
1. Our email notifications should be migrated and kept intact during the update process
2. I think there's a big problem with forcing email notifications to create actual template files on a version controlled site, and allowing users to create notifications the same way on the production server. I'm expecting our deployments to start failing once there are merge conflicts. These should be kept in the database to prevent this.

**Craft & Plugin Info (please complete the following information):**

- Craft Version: 4.0.5
- Freeform Version: 4.0.0-beta12
- Freeform Edition: Pro
- Fresh Install or Upgrade: Upgrade

**Comment by kjmartens** [comment]
Hi @darylknight,

The Freeform 4 upgrade removes the option to use database-based email notifications, as we've moved on to a unified experience:
https://docs.solspace.com/craft/freeform/v4/setup/updating-freeform-3.html

Freeform 3.11 added the ability to create and update existing file-based email notification templates inside the control panel, making the database-based email notification templates redundant. Since the experience is identical, if you wish to create and manage email notification templates inside the control panel, you can continue to do so, but they're stored as files instead of in the database. 🙂 

Also, we left the migration utility for anyone that forgot to convert their database templates to file templates. You should be able to do this inside the Email Templates Settings area of the control panel. 🙂 

In terms of preventing client users from creating or editing email notification templates, this can be done by removing the **Access Email Templates** and/or **Manage Email Templates** permission from the user/user group.

**Comment by darylknight** [comment]
Thanks for clarifying Kelsey. In general, I feel like I'm pretty good at paying attention to patch notes, and I wasn't aware of any of these changes. Can I suggest that one of those orange boxes is added somewhere to the Updates section (inside Craft's plugin changelog, not inside Freeform) which tells people to look at this? Losing all the email templates is going to catch people out, and I was completely unaware of the changes to move notifications into files.

Could you please explain how this is going to fit in with version controlled projects? If Freeform is going to use files as notification templates now, I see that running into merge conflicts when we try to deploy. I created two notifications locally. That created two files which I've now checked into git - but they are going to be overwritten on the server if anyone makes changes to them in the control panel. Next time I deploy, or edit them locally, the deployment will fail because the files on the server have changed.
I can't exclude the email templates folder, because otherwise the live server will be expecting to find a template file that isn't there if it was created locally. Revoking client permissions to edit notifications isn't really ideal as we'd be removing something they're already using.

I guess it's too late now as it's already built, but just sharing my opinion that I much preferred being able to do this in the database, as they're user editable content - not structure.

**Reply by michtio** [reply]
As follow-up on what @darylknight just mentioned. You create files for notifications... In my honest opinion a plugin should never ever write files on the server. Version Control deployments will break/fail and each time you deploy according to atomic design principles the templates created on server will disappear. 

Symlinking is a possibility, but that means uploading templates by hand.

**Reply by kjmartens** [reply]
Thanks for your feedback @darylknight and @michtio. I suppose it depends on the exact flow for your site development. It's entirely possible to prevent users from accessing/creating/editing email notification templates from the control panel via Craft user/group permissions. Is that not an option? Or is it that they are Admin accounts and the permissions don't apply? If so, perhaps another environment-specific setting to disable this area from showing for admins as well? Or do you require that your client users still create/edit email notifications on the production environment? 🙂 

We will review this and give it more thought. It's not too late for us to "undo" this change, as the email notification templates that are stored in the database are still there after v4.0 migration.

As for the warning notice in the Craft control panel, this was included in the changelog as a warning message and would have appeared as an orange box. 🙂 
https://github.com/solspace/craft-freeform/blob/v4.0/master/CHANGELOG.md#400-beta1---2022-03-29

**Comment by darylknight** [comment]
Off the top of my head, these are the reasons it's not great to just prevent clients from creating email notifications anymore:
1. They've been able to create and edit email notifications in Craft 3
2. They're content. They should be editable by clients.
3. The client can create new forms in the control panel. They should be able to create their own notifications for those new forms through the control panel too
4. If we turn off the ability for users to create and edit notifications, they'll just end up asking us to do it for them, which we'll probably forget to do locally if the option is still visible on the production site, and break deployments again

I think it's safe to assume that most developers will be using version control and project config (and that's the recommended Craft way to do things), and therefore plugins shouldn't be creating files on the live server under any circumstances. Doesn't matter whether it's atomic or just normal version control - our non-atomic deployment will fail too because version controlled files on the server have changed. If we don't version control them (by ignoring `templates/_freeform`), Freeform will fall over because the template file it's looking for doesn't exist unless we transfer them to the server manually.

------

Ah, so the reason I've never seen that orange box in the changelog is because the latest version on Craft 3 is 3.13.13. When you update to Craft 4, you use the "Craft 4 Upgrade" tool and see that the latest version of Freeform is 4.0.0-beta.12, and you install that. So you never see the changelog for 4.0.0-beta.1.

**Reply by kjmartens** [reply]
Ok, fair enough. We'll discuss this internally and very possibly add back support for this. 🙂

**Reply by darylknight** [reply]
Thanks Kelsey :) Please do let me know what you decide either way - this'll be holding us back from updating sites to Craft 4 and , without wanting to sound overly-dramatic, may stop us from using Freeform on new builds if they stay as file-only.

**Reply by kjmartens** [reply]
We're working on "undoing" this change now. Thanks again for your feedback. 🙂

**Reply by leigeber** [reply]
That's great news @kjmartens. If you can't say not worries, but is there an ETA on that? We have something blocked by this and need to plan accordingly. Thanks.

**Reply by kjmartens** [reply]
@leigeber We'll likely have something by early next week. 🙂

**Reply by darylknight** [reply]
Excellent, thanks for the update!

**Reply by darylknight** [reply]
@kjmartens Any news?

**Reply by kjmartens** [reply]
@darylknight the feature took a bit longer than expected but has been readded (and a bit cleaner now too). However, there are a couple of bugs we have to fix with it. It should be available tomorrow or Thursday. 🙂

**Comment by kjmartens** [comment, marked as answer]
We have now reversed the decision to remove database notifications. You can now continue to create and use email notification templates stored as files, in the database, or a combination of both as of Freeform 4.0.0-beta.14, now available (July 12, 2020).

Thank you again for your feedback on this. We are always listening while innovating Freeform, and this was one aspect of the change that we overlooked. We're sincerely sorry for the inconvenience caused. 🙂

**Reply by darylknight** [reply]
Thank you for listening Kelsey! Really appreciate it.

**Comment by darylknight** [comment]
@kjmartens I think this can be removed from the Preflight checks now?

![image](https://user-images.githubusercontent.com/1250017/180742548-8cf6e969-a1cf-446c-a806-815de2b94457.png)

**Reply by kjmartens** [reply]
Yes, this has been removed already for the next release. 🙂
